src/cgutils.cpp

```diff
@@ -919,6 +919,56 @@ static void raise_exception_if(Value *cond, Value *exc, jl_codectx_t *ctx)
                            exc, ctx);
 }
 
+static size_t dereferenceable_size(jl_value_t *jt) {
+    size_t size = 0;
+    if (jl_is_array_type(jt)) {
+        // Array has at least this much data
+        size = sizeof(jl_array_t);
+    } else {
+        size = jl_datatype_size(jt);
+    }
+    return size;
+}
+
+static inline void maybe_mark_argument_dereferenceable(Argument *A, jl_value_t *jt) {
+    if (!jl_is_leaf_type(jt)) {
+        return;
+    }
+    size_t size = dereferenceable_size(jt);
+    if (!size) {
+        return;
+    }
+    llvm::AttrBuilder Attrs;
+    Attrs.addDereferenceableAttr(size);
+#if JL_LLVM_VERSION >= 50000
+    A->addAttr(llvm::AttributeList::get(jl_LLVMContext,
+                                        A->getArgNo() + 1, Attrs));
+#else
+    A->addAttr(llvm::AttributeSet::get(jl_LLVMContext,
+                                        A->getArgNo() + 1, Attrs));
+#endif
```

Comment:
You should tell the clang folks about that because they do it all over the place, e.g. https://github.com/llvm-mirror/clang/blob/master/lib/CodeGen/CGCall.cpp#L2315-L2319